File: src/main/java/com/twsg/lnl/kotlin/c8_singleton/g1_vanilla_eager/WogaaEngine.java
package com.twsg.lnl.kotlin.c8_singleton.g1_vanilla_eager;

/**
 * @author tonyl
 */
public class WogaaEngine {

    public static void setupEnvironment() {
        System.out.println("Setting up environment");
    }

    public static final WogaaEngine INSTANCE = new WogaaEngine();

    private WogaaEngine() {
        System.out.println("Eager initialization");
    }

    public void analyze() {
        System.out.println("Analyzing");
    }

}

File: src/main/java/com/twsg/lnl/kotlin/c8_singleton/g3_kotlin/WogaaEngine.kt
package com.twsg.lnl.kotlin.c8_singleton.g3_kotlin

/**
 * @author tonyl
 */
object WogaaEngine {

    init {
        println("Heavy initialization")
    }

    fun analyze() {
        println("Analyzing")
    }

    @JvmStatic
    fun setupEnvironment() {
        println("Setting up environment")
    }
}
